Bug 1994457: Update to kubernetes 1.21.4

DhritiShikhar opened the pull request:
#### Special notes for your reviewer:

Conflicts can be verified in 8b40b8732fe6f7b767cf2f4630fd74384228761a

#### Git log

```
286709307da (tag: v1.21.2-rc.0) Release commit for Kubernetes v1.21.2-rc.0
14df10f60bb Update CHANGELOG/CHANGELOG-1.21.md for v1.21.1
b047b4c23ad Make watch order conformance test reliable
f204fd30d0d Ref counting is only applicable to Remote endpoints
91ce2b1f7f1 Azure: avoid setting cached Sku when updating VMSS and VMSS instances
cd34e81abae fix: delete non existing disk issue
dc12c607582 kubelet: reuse manager
578d22ab644 pkg/kubelet/nodeshutdown/systemd: fix for dbus 5.0.4
5eea5478ae4 vendor: bump runc to rc95
37d50b825dd Set cgroups via opencontainer
f283c607428 Fix expired unit test certs
02e92912478 Remove unnecessary quotes from get-kube scripts
6732b634957 Respect annotation size limit for SSA last-applied.
8717fcf8bec Upgrade konnectivity-client for GRPC connection fixes
d2e8584ff19 Update debian-iptables to buster-v1.6.1
72600375068 Update debian-base to buster-v1.7.0
98e8c648fe5 Update etcd image revision
d5c699de51d kubeadm: remove e2e test for ClusterStatus
c3d8507df1b Update setcap to buster-v2.0.1 and add setcap to dependencies.yaml
539bb658531 fix error of setting negative value for containerLogMaxSize
438f0d695c2 Fix VolumeAttachment garbage collection for migrated PVs
ae40fd73208 (scheduler e2e) Create balanced pods in parallel
38a41e15576 use subpath for coredns only for default repository
19e8f2bab0c Return UnschedulableAndUnresolvable instead of Error when failing to lookup pvc or storageclass in VolumeZone plugin
a62c5dc20a2 Return UnschedulableAndUnresolvable when looking up volume-related resources returns NotFound error
b0daa8fa4ea serviceOwnsFrontendIP shouldn't report error when the public IP doesn't match
46b90a00e5d Fix closing of decorated watcher channel on timeout
154a5538a5c Revert "Cleanup portforward streams after their usage"
4977a4c2085 sched: fix a bug that a preemptor pod exists as a phantom
99b213bba67 endpointslicemirroring controller mirror address status
4033f7a5691 feat: remove ephemeral-storage etcd requirement
ddf5532facb Update to go1.16.5
ef0beb2cf70 client-go: reduce log level of reflector again
2cf51ac8b12 Remove error wrap from logs
f7e8a1ec257 strip prefix 'v' in coredns version if exists
092fbfbf534 (tag: v1.21.2) Release commit for Kubernetes v1.21.2
ad0c897c883 (tag: v1.21.3-rc.0) Release commit for Kubernetes v1.21.3-rc.0
3aeb38302e5 Update CHANGELOG/CHANGELOG-1.21.md for v1.21.2
9eac5b50dda Do not throw error when we can't get canonical path
22d66e792d7 fix: cleanup outdated routes
afa278757d2 fix: Refresh VM cache when node is not found
2cb3e1de2e5 Fix Node Resources plugins score when there are pods with no requests
2fdb62fcdf8 Fix frameworkImpl.extenders being not set
3e12efadd4e Fix race in attachdetach tests
6effb43f755 Loadbalancer IngressIP policy should be configured as non-DSR to enable routing mesh by default
318677cc93e generate scheduler merge patches on the pod status instead of the full pod
44d4c4fe69f Bump SMD to v4.1.2 to pick up #102749 fix
aac1ab5152b Remove extra zone test
e07463492f5 wait for endpoints to be available
a47bc4e390a fix: return empty VMAS name if using standalone VM
d4da53771f4 Simplify use of the fake dynamic client
73bbff81403 Update to go1.16.6
60ebb8b667a Update debian-base image to buster-v1.8.0
2d6dd4e284e Update debian-iptables image to buster-v1.6.5
0ce3bf060e2 Update setcap image to buster-v2.0.3
8ceaecb07d4 Fix the code is leaking the defaulting between unrelated pod instances
8853e418c10 Updated to use konnectivity client v0.0.21, and implemented placeholder context
131ff61bf1 move upgrade test frameworks closer to Describe
ca643a4d1f7 (tag: v1.21.3) Release commit for Kubernetes v1.21.3
ff7c569f157 (tag: v1.21.4-rc.0) Release commit for Kubernetes v1.21.4-rc.0
1c29b7706e7 Update CHANGELOG/CHANGELOG-1.21.md for v1.21.3
01da25eaef1 Fix panic in master upgrade tests
41d51d2daf6 kubelet: do not call RemoveAll on volumes directory for orphaned pods
e62a8123296 update comments and owners file for pkg/util/removeall
a7a15f4d851 storage e2e: automate hostpath YAML updates, hostpath v1.6.2
d2fa4479bc8 storage e2e: disable health check containers
a9543fbfdfe storage e2e: downgrade hostpath driver
38ec624c919 storage e2e: patch in RBAC rules for secrets
5b5655c48d7 Update pd csi driver images to use v1 images (CSINode, CSIDriver, etc)
c6304d0287c disable aufs module
3ab6a85abe7 Make CSR cleaner tolerate objects with invalid status.certificate
21836789d33 Update to using apiserver-network-proxy v1.22
4f462a0f6d9 feat: Provide IPv6 support for internal load balancer
0ac9e77a292 Fix: ignore not a VMSS error for VMAS nodes in reconcileBackendPools
c37bf8c3cf1 Fix disruptive subPath test failures
961a35ed44d Update configure-helper.sh
88f2a255530 Update configure-helper.sh
caac1b7fcf8 Update configure-helper.sh
77dcc9b53c7 Fix metrics reporting for the deprecated watch path
3ab23b9f198 Update to go1.16.7
98d37747d8f Avoid spurious calls to update/delete validation
3cce4a82b44 (tag: v1.21.4) Release commit for Kubernetes v1.21.4
```

DhritiShikhar commented:
/test images

DhritiShikhar commented:
> is this correct? The lower block is for what upstream did, i.e. bumping to `structured-merge-diff/v4 v4.1.2`. Shouldn't we follow that here and every other similar case (where we don't have forks) ?

@sttts We are using `structured-merge-diff/v4 v4.1.2` -> https://github.com/openshift/kubernetes/pull/888/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R137

DhritiShikhar commented:
/test e2e-aws-fips

soltysh commented:
> @sttts We are using `structured-merge-diff/v4 v4.1.2` -> https://github.com/openshift/kubernetes/pull/888/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R137

@DhritiShikhar I think what Stefan is saying that even though we are using the correct version in replace directive, the conflict resolution commit contains wrong fix for that part. I also noticed we're using wrong `k8s.io/utils` version, see https://github.com/kubernetes/kubernetes/blob/v1.21.4/go.mod#L139-L140 You'll need to fix that.

soltysh commented:
Overall this looks good aside from my previous comment.

DhritiShikhar commented:
@soltysh Addressed your comments. 

Updated `structured-merge-diff` to `v4.1.2` version and `k8s.io/utils` to `v0.0.0-20201110183641-67b214c5f920` version.

tjungblu commented:
/retest-required

tjungblu commented:
/bugzilla refresh

DhritiShikhar commented:
/bugzilla refresh

tjungblu commented:
/retest-required

tjungblu commented:
/bugzilla refresh

tjungblu commented:
/bugzilla refresh

soltysh commented:
/override ci/prow/verify-commits
that will never pass on a k8s update.

soltysh commented:
/retest

soltysh commented:
@DhritiShikhar looks like during rebasing this PR you broke the commits, I'd expect the following:

1. merge commits, which will be something like `Merge tag 'v1.21.4' into rebase-1.21.4`
2. conflict resolution commit `UPSTREAM: <drop>: manually resolve conflicts`
3. all the remaining updates `UPSTREAM: <drop>: hack/update-vendor.sh, make update and update image`

see last 4 commits in https://github.com/openshift/kubernetes/pull/916/commits, for example. In your PR it looks like the first commit (`UPSTREAM: <drop>: manually resolve conflicts`) is confusing because it's not conflict resolution but a merge from what I see. Then you have `UPSTREAM: <drop>: hack/update-vendor.sh` which is actually conflict resolution, and the last one which is correct. It's important to keep the history clean and understandable so you'd need to fix that before merging it. 

Also as discussed earlier today please update the description of https://bugzilla.redhat.com/show_bug.cgi?id=1994457 with information about changes from 1.21.1 up to 1.21.4.

DhritiShikhar commented:
/retest-required

soltysh commented:
Manually adding valid-bug label, based on discussion in control plane arch call.

DhritiShikhar commented:
/bugzilla refresh

soltysh commented:
/override ci/prow/verify-commits
that will never pass on a k8s update.

soltysh commented on go.mod:
```diff
@@ -196,7 +197,7 @@ replace (
 	github.com/chzyer/readline => github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e
 	github.com/chzyer/test => github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1
-	github.com/cilium/ebpf => github.com/cilium/ebpf v0.6.2
+	github.com/cilium/ebpf => github.com/cilium/ebpf v0.5.0
```
This seems wrong, we shouldn't be changing this.

soltysh commented on go.mod:
```diff
@@ -215,7 +216,7 @@ replace (
 	github.com/coreos/go-semver => github.com/coreos/go-semver v0.3.0
 	github.com/coreos/go-systemd => github.com/coreos/go-systemd v0.0.0-20190321100706-95778dfbb74e
-	github.com/coreos/go-systemd/v22 => github.com/coreos/go-systemd/v22 v22.3.2
+	github.com/coreos/go-systemd/v22 => github.com/coreos/go-systemd/v22 v22.3.1
```
Same here.

soltysh commented on go.mod:
```diff
@@ -284,8 +285,8 @@ replace (
 	github.com/google/btree => github.com/google/btree v1.0.0
-	github.com/google/cadvisor => github.com/openshift/google-cadvisor v0.33.2-0.20210610135131-57b941c7657a
-	github.com/google/go-cmp => github.com/google/go-cmp v0.5.5
+	github.com/google/cadvisor => github.com/google/cadvisor v0.39.0
```
Same here.

DhritiShikhar commented:
/test e2e-aws

vrutkovs commented:
/retest

soltysh commented:
/override ci/prow/verify-commits
that will never pass on a k8s update.

/test e2e-metal-ipi
/test e2e-metal-ipi-ovn-ipv6
/test e2e-metal-ipi-ovn-dualstack

soltysh reviewed (approved):
/lgtm
/approve

DhritiShikhar commented:
/test e2e-metal-ipi-ovn-dualstack

sdodson commented:
/bugzilla refresh

hardys commented:
[patch-manager] Approving cherry pick with score 1.0

soltysh commented:
/override ci/prow/verify-commits
that will never pass on a k8s update.

/retest-required
